### ISSUE TITLE
[FEATURE] Metadata schema

### ISSUE THREAD

**Issue by sdatkinson** [sdatkinson/neural-amp-modeler#176]
This Issue is for defining and implementing a schema for metadata.

* `"display_name"` (`str`) What the model name might be displayed as (instead of the file name!)
* `"loudness"` (`float`, <= 0.0) The output (in dB) of a model, given a nominal signal. (Implemented in #137)
* `"gear_type"` (`enum`): `"amp"`, `"pedal"`, `"amp_pedal"`.
* `"with_cab"` (`bool`)
* `"compression"` (`float`, between 0.0 and 1.0) a quantification of how much compression the model imparts on a signal. Compression of 0.0 means that the response is linear in the input; 1.0 means that an infinitesimal input signal would cause the model to output a signal at its max loudness (example: `y = a * sign(x)`)
* `"character"` (`enum`): `"clean"`, `"crunch"`, `"hi_gain"`
* `"tags"`: a `list` of `str`s
* `"modeled_by"`: (`str`) the name of the person who made the model
* `"date"`: `{"year": int, "month": int, "day": int}` when the model was exported to a .nam file (e.g. determined during [`.export()`](https://github.com/sdatkinson/neural-amp-modeler/blob/main/nam/models/_exportable.py#L24).
* `"instrument"`: (`enum`) `"bass"`, `"guitar"`, `"other"` what instrument the model is intended to be used with. (wrt `"other"`, there have been instances of e.g. modeling amps intended for harmonicas!)
* `"misc"`: (`str`) Anything else?
* `"standard_architecture"` (`enum`) `"standard"`, `"lite"`, `"feather"` (if the architecture matches one of the architectures I've made easy to pick), or `None`.
* `"esr_val1"`: (`float` or `None`) The final ESR, if the model was validated on the validation set from the `v1_1_1.wav` file (If you used different training data, then the ESR wouldn't be "apples to apples". There's also the possibility of validation set leakage...). So long as the validation st stays the same, this can continue to be reported. This woudl be retired for e.g. `"esr_val2"` if a future standard training signal changed the validation set.

I'd like for there to be metadata about the gear being modeled, but I'm not sure what the most reasonable way to do this is.

Option "1":
* `"gear_modeled"`: (`str`) Just write it out
Option "2": "make" and "model" (and "year"?)
Option "n": 
* `"gear_modeled"`: a full-on graph where each node is an element of the signal chain. (Yikes...can you imagine?)

Seeking suggestions and feedback, esp. about how to represent the gear being modeled.

Edits: 
* 2023-04-08 4:15p Add `"standard_architecture"`, `"esr_val1"`

**Comment by heebje** [comment]
- Gear type could have more entries (for tape machines, etc). Maybe something like "studio_hardware"?
- Licence! The licence that the creator of the profile wants to attach to it, if any. Like, is it public domain, or CC, or commercial, or ...?

**Comment by mikeoliphant** [comment]
"final_esr" - The final esr of the model at the end of training.
"sample_rate" - The sample rate used for training.

**Comment by gtirard** [comment]
Could we include an image of the orginal amp in the model's metadata?

**Comment by JZ1978** [comment]
Not sure if this would make it easy for people to "steal" other peoples work and put new/fake metadata on other peoples models but.... It would be nice if there was a way to write metadata onto the already created .nam files. So the last few months of contributions can have the same search features as any new stuff that comes out once this is implemented.

**Comment by colin-campbell** [comment]
Essential: a metadata format "version".
Awesome: A license string, A bag of parameter names, types? and values (that UIs can interpret how they wish), a automatic cryptographic MAC hash for the capture data (and/or) metadata (authentication), and that hash signed with a cryptographic sig of the author (Attribution).

**Comment by Aelfstone** [comment]
"gear_type" could also be "preamp" and "other".

**Comment by sdatkinson** [comment]
@heebje 
>     * Licence! The licence that the creator of the profile wants to attach to it, if any. Like, is it public domain, or CC, or commercial, or ...?

Is there precedence for holding this as metadata in a file?

**Comment by sdatkinson** [comment]
@mikeoliphant 
> "sample_rate" - The sample rate used for training.

I intend that to a a proper part of the model (not just metadata) when non-48k is supported in the trainer (see #13)

**Comment by sdatkinson** [comment]
@gtirard 
> Could we include an image of the orginal amp in the model's metadata?

I'm not keen on that for now, since I'm basically using the JSON file format for the moment 🙂

**Comment by sdatkinson** [comment]
@JZ1978 
> Not sure if this would make it easy for people to "steal" other peoples work and put new/fake metadata on other peoples models but.... It would be nice if there was a way to write metadata onto the already created .nam files. So the last few months of contributions can have the same search features as any new stuff that comes out once this is implemented.

I wouldn't be against it. I don't think this makes it meaningfully harder to "steal" anyone's models; it's mainly to help facilitate building displays, databases, and such.

**Comment by sdatkinson** [comment]
@colin-campbell 
> Essential: a metadata format "version". Awesome: A license string, A bag of parameter names, types? and values (that UIs can interpret how they wish), a automatic cryptographic MAC hash for the capture data (and/or) metadata (authentication), and that hash signed with a cryptographic sig of the author (Attribution).

I'm interested in the hash, but not convinced that it does its job. Why couldn't I just rewrite the hash with my own? (Infosec isn't my area of expertise.)

**Comment by sdatkinson** [comment]
@scottcorgan
> On of the only things that matters from the ToneHunt perspective of parsing the is that it remains valid JSON. 

I won't commit to that (hence why I renamed the extension to .nam instead of .json), but if it _does_ become incompatible w/ the JSON spec, I'll give you tools to load the model 👍🏻 (Using Python, in this repo; C++, in [NeuralAmpModelerCore](https://github.com/sdatkinson/NeuralAmpModelerCore)). And I'll publish a spec for the file format.

> Other than that, a strategy to make sure there is versioning in place that will help us parse updates (various versions) to the .nam file spec as it progresses.

The plan is to handle this with the main model version, which I've fixed to the repo version (current `main` branch is 0.5.1) and the intent is to follow semantic versioning. Since this is pre-v1 and it seems (to me) that there's a little bit of ambiguity on it, I'm bumping the minor version when breaking changes are introduced; patch otherwise. So adding new fields would bump the patch version; renaming or removing one would minor. Open to feedback on this, but I _think_ it'll be fine.

> We'll track our work in our own issue when it's ready [scottcorgan/tonehunt#161](https://github.com/scottcorgan/tonehunt/issues/161)

🎉

**Comment by rossbalch** [comment]
I like the current set you have going on there, I think it's important not to go too over board. I think option two for the gear modelled is good. Make people be concise. "Manufacturer", "Model", "Channel"

I can't think of any other fields I would like to see. Maybe "Direct" vs "Speaker"

I my opinion any other details can be described by a user as they share the file.

**Comment by fichl** [comment]
Hi everybody,
just an idea... how about putting gear specific frequencies for bass, mid and treble in the metadata.
maybe a model of a bassamp would rather eq @ 100hz instead of 150hz.
if left blank, defaultsettings are used.

**Comment by colin-campbell** [comment]
> I'm interested in the hash, but not convinced that it does its job. Why couldn't I just rewrite the hash with my own? (Infosec isn't my area of expertise.)

@sdatkinson  In short:
* The MAC hash authenticates (identifies) the contents of the "message", ie the integrity of the nam file - any app will know if a nam file has been tampered with or corrupted in some way. It doesn't provide secure attribution of authorship - although If for example you hash the metadata and that metadata includes for example a string "© 2023-04-09 Steven Atkinson", you will know if your work is being ripped off if someone changes it and passes it off as their own, because the hash won't match your hash.

* The secure attribution (and non-repudiation) comes from digitally signing your hash with a private key and including the public key in the file. This ensures that people can't impersonate you.  There are existing technologies like JWS https://en.wikipedia.org/wiki/JSON_Web_Signature that take care of the technical details. However what these don't do on their own establish **trust**.

* Trust is established by verification of the identity of the publisher, and that is the tricky part - once could use an ssh key attached to a Github account, for example, but that is not always great for mere mortals. However the good news is that there are open platforms for this like https://www.sigstore.dev (like letsencrypt for code) that help with this.

I'm happy to discuss this further with you. My take on this - as with the licensing question in the other issue - is ensuring that Free Software creators don't get ripped off of their works, and not allowing proprietary commercial interests an unfair advantage by ripping these people off.

**Comment by gtirard** [comment]
I don't if it's the right place to post here, sorry if it isn't. Also don't know if anybody has already discussed this…

When the user captures the same amp with different settings, it would be nice to be able to load a group of models with virtual knobs corresponding to the amp's setting. Instead of loading each model one by one, trying to guess the original amp's setting by reading the name of each model, the end user would select option in the UI, and it would load the corresponding model.

Suppose we have these models:

![image](https://user-images.githubusercontent.com/668338/231982457-5c0f01ad-57a6-48c9-a51c-60dab886e352.png)

We would create a JSON conf file, to link these models together, and be able to show them a different setting of one amp, like this:

![image](https://user-images.githubusercontent.com/668338/231982739-18e8b53b-6f73-4513-92d5-9cb0a5bda504.png)

**Comment by sdatkinson** [comment]
Thanks, all!

Closing this as the first version of the metadata schema is shipped in v0.5.1.

Let's give it a spin and re-visit if anything is obviously lacking. If so, then new issues may reference topics discussed here.
